src/codegen/strength_reduce/reaching_values.rs

```diff
@@ -0,0 +1,207 @@
+    block_no: usize,
+    cfg: &ControlFlowGraph,
+    vars: &mut Variables,
+    block_vars: &mut HashMap<usize, Variables>,
+    ns: &Namespace,
+) {
+    // We should merge the incoming set of variables with the existing ones. If there
+    // are no changes, then we cease traversing the cfg. The rules are:
+    // - If there are more than MAX_VALUES entries in the result, make the set the unknown set
+    // - If either the existing set or the incoming set contains unknown, make set the unknown set
+    // - If there are no changes to the existing set, record this
+    // - This is a very hot code path. This needs to be _FAST_ else compilation time quickly explodes
+    if let Some(map) = block_vars.get_mut(&block_no) {
+        let mut changes = false;
+
+        for (var_no, set) in vars.iter() {
+            changes |= update_set(var_no, set, map);
+        }
+
+        if !changes {
+            // no need to do this again
+            return;
+        }
+    } else {
+        block_vars.insert(block_no, vars.clone());
+    }
+
+    for instr in &cfg.blocks[block_no].instr {
+        transfer(instr, vars, ns);
+
+        match instr {
+            Instr::Branch { block } => {
+                // must be last in the block
+                reaching_values(*block, cfg, vars, block_vars, ns);
+            }
+            Instr::BranchCond {
+                cond,
+                true_block,
+                false_block,
+            } => {
+                // must be last in the block
+                let v = expression_values(cond, vars, ns);
+
+                if v.len() == 1 {
+                    let v = v.iter().next().unwrap();
+
+                    // if we know the value of the condition, follow that path
+                    if v.known_bits[0] {
+                        reaching_values(
+                            if v.value[0] {
+                                *true_block
+                            } else {
+                                *false_block
+                            },
+                            cfg,
+                            vars,
+                            block_vars,
+                            ns,
+                        );
+
+                        continue;
+                    }
+                }
+
+                // we don't know the value of the condition. Follow both paths
+                let mut vars_copy = vars.clone();
+
+                reaching_values(*true_block, cfg, &mut vars_copy, block_vars, ns);
+
+                reaching_values(*false_block, cfg, vars, block_vars, ns);
+            }
+            Instr::AbiDecode {
+                exception_block: Some(block),
+                ..
+            } => {
+                let mut vars = vars.clone();
+
+                reaching_values(*block, cfg, &mut vars, block_vars, ns);
+            }
+
+            _ => (),
+        }
+    }
+}
+
+fn update_set(var_no: &usize, set: &HashSet<Value>, map: &mut Variables) -> bool {
```

Comment:
This function looks odd to me. I don't like the idea of passing a mutable map and returning a bool to inform the user whether the map was mutated. I'd rather not change the map in place, instead returning the HashSet containing prospect changes to the user. The user can get to his bool then for free anyways if he really needs it:  `update_set(...).is_empty()`. (If you really do need to have in place modification, I think a doc comment on whats happening exactly would be very helpful, also in this case the function name "update_set" can be very misleading when in fact the passed `map` and not the `set` get changed).

I allowed myself to refactor this function as follows making it shorter and in my opinion much easier to grasp whats going on: 

```rust
fn update_set(var_no: &usize, set: &HashSet<Value>, map: &Variables) -> HashSet<Value> {
    if let Some(existing) = map.get(var_no) {
        if existing.iter().next().map_or(false, |v| v.all_unknown()) {
            return HashSet::new();
        } else if let Some(v) = set.iter().find(|v| v.all_unknown()) {
            // If we are merging an unknown value, set the entire value set to unknown
            return HashSet::from_iter([v.clone()]);
        } else {
            let filtered = set.iter().filter(|v| existing.contains(v)).cloned();
            let mut changes = HashSet::from_iter(filtered);
            if existing.len() + changes.len() > MAX_VALUES {
                let bits = existing.iter().next().unwrap().bits;
                changes.insert(Value::unknown(bits));
                return changes;
            }
            return HashSet::new(); /* We dont return our changes here, is this intended? */
        }
    };

But please beware I could have made mistakes as well, above code snipped compiles but I didn't check for correctness otherwise.
    // We have no existing set. Create one but folding unknown
    if set.len() > MAX_VALUES || set.iter().any(|v| v.all_unknown()) {
        let bits = set.iter().next().unwrap().bits;
        return HashSet::from_iter([Value::unknown(bits)]);
    }
    set.clone()
}
```

Reply:
This function is performance critical, we should reduce allocation because they totally kill performance. The strength reduce pass is often the most cpu intensive pass. Doing `.clone()` for each variable mutation will be a disaster.

Secondly the function you proposed is not doing the right thing. The variable `map: Variables` should be updated.

Reply:
The clones are exactly the same ones as in the original code? I was thinking about return a HashSet containing refs which would avoid cloning entirely, but as far as I understood they would need to be cloned anyways for inserting them into `Variables`.
Returning `HashSet::new()` should not pose any overhead as well as it doesn't allocate.

> Secondly the function you proposed is not doing the right thing. The variable map: Variables should be updated.

The idea of the refactored code is that the caller would extend his Variables map himself. We can leave the function as it is if it works its fine I guess. But I'd at least change the name so it would imply the `map` is changed (and not the set), I was quite confused by that...

Reply:
The function `reaching_values` is called 1300 times in a recursive manner in an example I found. The purpose of the `update_set` function is to decrease the former's stack size without affecting performance. 

> The idea of the refactored code is that the caller would extend his Variables map himself.

This is going to add another overhead to our function. Returning the hash set not only will increase stack size, but also will need another clone when extending the original set.

Reply:
> Returning the hash set not only will increase stack size,

Agree on that

> but also will need another clone when extending the original set.

Why?

Reply:
I still fail to see where my refactored code snippet adds any more clones/allocations that weren't already there. For example at Line 102 in the original code:

```rust
let mut set = HashSet::new();

set.insert(v.clone()); 

changes = true;
map.insert(*var_no, set);
```
Unless I'm missing something here, this already triggers 3 memory allocations, doesn't it? Or how is this different from `return HashSet::from_iter([v.clone()]);` and extending the map using that? Extending the map could even be faster in some cases here because you know before hand how much additional memory to allocate (vs. having multiple `insert` calls).  

I'm sure this could be refactored to use less allocations.. But is this really a problem for now? If it was fine before, just change the function name to something less confusing and maybe add a comment so it is clear what's happening and I'm happy with that.

Reply:
@xermicus after re-examining this I see your point. How about going one step further do something like this:

```rust
fn update_set(set: &HashSet<Value>, existing: &Option<HashSet>) -> HashSet<Value> {
     if let Some(existing) = existing {
        if existing.iter().next().map_or(false, |v| v.all_unknown()) {
            HashSet::new()
        } else if let Some(v) = set.iter().find(|v| v.all_unknown()) {
            // If we are merging an unknown value, set the entire value set to unknown
            HashSet::from_iter([v.clone()])
        } else {
            let filtered = set.iter().filter(|v| existing.contains(v)).cloned();
            let mut changes = HashSet::from_iter(filtered);
            if existing.len() + changes.len() > MAX_VALUES {
                let bits = existing.iter().next().unwrap().bits;
                changes.insert(Value::unknown(bits));
                changes
            } else {
               HashSet::new() /* We dont return our changes here, is this intended? */
            }
        }
    // We have no existing set. Create one but folding unknown
    } else if set.len() > MAX_VALUES || set.iter().any(|v| v.all_unknown()) {
        let bits = set.iter().next().unwrap().bits;
        HashSet::from_iter([Value::unknown(bits)])
    } else {
        set.clone()
    }
}
```
Now that the `map` has been removed completely this function is much clearer.

This refactor does rely on `HashSet::new()` not allocating, which is news to me. How does that work?

Thanks for a good idea @xermicus

Reply:
@seanyoung if we are going for this please check again my `// XXX` comments

Reply:
> Also if the stack size is a problem returning empty HashSet is a problem we could try to inline this function?

Inlining the function will only add more instructions to `reaching_values`, increasing its stack size even more.

> this sounds like super premature optimization,

Without these changes, strength reduce overflows its stack.

Reply:
```solidity
            let filtered = set.iter().filter(|v| existing.contains(v)).cloned();
            let mut changes = HashSet::from_iter(filtered);
            if existing.len() + changes.len() > MAX_VALUES {
                 // XXX should be like that? 
                let bits = existing.iter().chain(changes.iter()).next().unwrap().bits; // I think we should be 
                changes.insert(Value::unknown(bits));
            } 
            changes  // XXX should be like that? 
```

We should add to `existing` the values of `set`. `set.iter().filter(|v| existing.contains(v)).cloned()` is the intersection between `existing` and `set`.

The update of map[var_no] happens in two different ways. When we crate a new set, we insert it in `map` (this overwrites any existing value). If we do not create a new set, we add the values of set to `existing`. Returning a HashSet would mean also discerning between when to insert a new value to `map[var_no]` and when to extend it.

Reply:
@xermicus I tried to use you suggestion, but I couldn't get it to work. Feel free to patch my branch to check if you can improve your suggestion. I have just modified the function's name and added a doc comment.

Reply:
Sure I can give it one more shot. I don't understand what `let bits = existing.iter().next().unwrap().bits;` does? I don't get why it chooses just whatever the `.next()` thing is inside `existing`. This HashSet has no specific ordering, or has it?

Reply:
As discussed, lets leave it as it is and address that in a different issue/PR.